handler.go

```diff
@@ -6,31 +6,85 @@ package main
 import (
 	"bytes"
 	"errors"
+	"github.com/chris-wood/ohttp-go"
+	"google.golang.org/protobuf/proto"
 	"io/ioutil"
+	"log"
 	"net/http"
 	"net/http/httputil"
-
-	"github.com/chris-wood/ohttp-go"
-	"google.golang.org/protobuf/proto"
+	"strconv"
 )
 
+// Description of the error handling in the specification:
+// https://ietf-wg-ohai.github.io/oblivious-http/draft-ietf-ohai-ohttp.html#name-errors:
+
+// 401 - Unauthorized in Gateway response
 var ConfigMismatchError = errors.New("Configuration mismatch")
+
+// 400 - BadRequest in Gateway response
 var EncapsulationError = errors.New("Encapsulation error")
-var TargetForbiddenError = errors.New("Target forbidden")
+
+// 400 - BadRequest in Payload response. Payload is not a valid protobuf or marshalling error.
+var PayloadMarshallingError = errors.New("Issues with payload marshalling (BHTTP or Protobuf)")
+
+// 403 - Forbidden in Payload response. The request is not allowed to be sent to the target.
+var GatewayTargetForbiddenError = errors.New("Target forbidden on gateway (request was blocked by gateway)")
+
+// 500 - Internal server error in Payload response. The request is failed to be processed after decapsulation.
+var GatewayInternalServerError = errors.New("The request is failed to be processed after decapsulation")
+
+// Errors happened during decapsulation/encapsulation are returned as gateway response's error status (401 and 400)
+func EncapsulationErrorToGatewayStatusCode(e error) int {
+	switch e {
+	case ConfigMismatchError:
+		return http.StatusUnauthorized
+	case EncapsulationError:
+		return http.StatusBadRequest
+	default:
+		return http.StatusBadRequest
+	}
+}
+
+// Errors happened after decapsulation are returned as encapsulated payload errors while gatewy status is 200
+func PayloadErrorToPayloadStatusCode(e error) int {
+	switch e {
+	case PayloadMarshallingError:
+		return http.StatusBadRequest
+	case GatewayTargetForbiddenError:
+		return http.StatusForbidden
+	case GatewayInternalServerError:
+		return http.StatusInternalServerError
+	default:
+		return 400
+	}
+}
+
+// unable to decapsulate the request or unable to encapsulate the response. Leads to 401 or 400 on gateway level
+func EncapsulationFail(err error) (ohttp.EncapsulatedResponse, error) {
+	return ohttp.EncapsulatedResponse{}, err
+}
+
+func EncapsulationSuccess(response ohttp.EncapsulatedResponse) (ohttp.EncapsulatedResponse, error) {
+	return response, nil
+}
+
+func PayloadSuccess(payload []byte) ([]byte, error) {
+	return payload, nil
+}
```

Comment:
```suggestion
```

I would remove these two functions. They don't seem to improve readability much at the cost of an extra jump.